invenio_vocabularies/datastreams/readers.py

```diff
@@ -87,4 +96,61 @@ def read(self):
                 # todo add logging/fail
                 pass
 
-            yield StreamEntry(resp.content)
+            yield resp.content
+
+
+class ZipReader(BaseReader):
+    """ZIP reader."""
+
+    def __init__(self, *args, options=None, regex=None, **kwargs):
+        """Constructor."""
+        self._options = options or {}
+        self._regex = re.compile(regex) if regex else None
+        super().__init__(*args, **kwargs)
+
+    def read(self, item=None):
+        """Opens a ZIP and iterates through the files in the archive."""
+        # https://docs.python.org/3/library/zipfile.html
+        filepath = item or self._origin
+        with zipfile.ZipFile(filepath, **self._options) as archive:
+            for member in archive.infolist():
+                match = not self._regex or self._regex.search(member.filename)
+                if not member.is_dir() and match:
+                    yield archive.open(member)
+
+
+class JsonReader(BaseReader):
+    """JSON object reader."""
+
+    def read(self, item=None, **kwargs):
+        """Reads (loads) a json object and yields its items."""
+        file = item if item else open(self._origin, mode='r')
```

Comment:
Is there a way maybe to use a `with ...` clause here (just to avoid the "manual" `file.close()` at the end). Also maybe we make a convention so that it's the responsibility of the previous reader to handle the open/close?

Reply:
there is but it would duplicate the inner code, will try to refactor.

Reply:
The base class implements now with the `with` ctx manager the opening of the file if it's not being piped through item. Some readers like Zip and Tar have to overwrite this.